File: src/readers/grib2/sections/7/templates.ts
import {
  JpxImage,
  complexUnpacking,
  imageDecoder,
  spectralComplexUnpacking,
  spectralSimpleUnpacking,
} from '../../../index.js';

import type { Grib2BitMapSection } from '../6/index.js';
import type { Grib2DataRepresentationSection } from '../5/index.js';
import type { Grib2Sections } from '../index.js';
import type { Reader } from '../../../index.js';

// TODO: spectrals MAY not be implemented correctly yet.
// TODO: case 41: case 40010: PNG
// TODO: case 42: AEC https://github.com/NOAA-EMC/NCEPLIBS-g2c/blob/develop/src/aecunpack.c

/**
 * Converts data Buffer according to data representation section
 * @param reader - The raw data to convert
 * @param sections - The sections of the GRIB2 message that have been parsed so far
 * @returns Converted data
 */
export async function getGrib2Template7(
  reader: Reader,
  sections: Grib2Sections,
): Promise<number[]> {
  const drs = sections.dataRepresentation;
  if (drs === undefined) throw new Error('Data Representation Section is not defined');
  const { dataRepresentationTemplate } = drs;

  switch (dataRepresentationTemplate.code) {
    case 0:
      return simpleUnpacking(reader, drs);
    case 2:
    case 3:
      return complexUnpacking(reader, sections);
    case 40:
    case 40000: {
      const bms = sections.bitMap;
      if (bms === undefined) throw new Error('Bit Map Section is not defined');
      return jpeg2000Unpacking(reader, drs, bms);
    }
    case 41:
    case 40010:
      return await pngUnpacking(reader, drs);
    case 50: {
      // 1. Unpack the "simple" part (indices 1 to ndpts-1)
      const unpackedData = spectralSimpleUnpacking(reader, drs);
      // 2. The first value (index 0) is often stored as a raw IEEE float in the DRS
      // In DRS 5.0/5.50, the referenceValue is that IEEE float.
      unpackedData.unshift(drs.dataRepresentation.referenceValue);

      return unpackedData;
    }
    case 51:
      return spectralComplexUnpacking(reader, sections);
    default:
      throw new Error(`Template 7.${dataRepresentationTemplate} not defined`);
  }
}

/**
 * # Data Template 7.0 - Grid point data - simple packing
 *
 * [Read more...](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_temp7-0.shtml)
 * @param reader - The raw data to convert
 * @param drs - The data representation section
 * @returns - The converted data
 */
export function simpleUnpacking(reader: Reader, drs: Grib2DataRepresentationSection): number[] {
  const { numberOfBits, decimalScaleFactor, referenceValue, binaryScaleFactor } =
    drs.dataRepresentation;

  const DD = Math.pow(10, decimalScaleFactor);
  const EE = Math.pow(2, binaryScaleFactor);
  const data = new Uint8Array(reader.slice().buffer);

  const totalBits = data.length * 8;
  const count = Math.floor(totalBits / numberOfBits);
  const values: number[] = [];

  let bitPos = 0;

  for (let i = 0; i < count; i++) {
    let acc = 0;

    for (let b = 0; b < numberOfBits; b++) {
      const byteIndex = Math.floor(bitPos / 8);
      const bitOffset = 7 - (bitPos % 8);
      const bit = (data[byteIndex] >> bitOffset) & 1;

      acc = (acc << 1) | bit;
      bitPos++;
    }

    values.push((referenceValue + acc * EE) / DD);
  }

  return values;
}

/**
 * # Data Template 7.40 - Grid point data - JPEG 2000 code stream format
 *
 * [Read more...](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_temp7-40.shtml)
 * @param reader - The raw data to convert
 * @param drs - The data representation section
 * @param bms - The bit map section
 * @returns - The converted data
 */
export function jpeg2000Unpacking(
  reader: Reader,
  drs: Grib2DataRepresentationSection,
  bms: Grib2BitMapSection,
): number[] {
  const jpx = new JpxImage(reader);
  if (jpx.componentsCount !== 1)
    throw new Error('JPEG Decoder: Only single component is supported');
  if (jpx.tiles.length !== 1) throw new Error('JPEG Decoder: Only single tile is supported');
  const pixels = jpx.tiles[0].items;

  const { bitMap: bitBuffer, bitMapIndicator } = bms;
  const { numberOfDataPoints, dataRepresentation } = drs;
  const { decimalScaleFactor, referenceValue, binaryScaleFactor, numberOfBits } =
    dataRepresentation;

  const DD = Math.pow(10, decimalScaleFactor);
  const EE = Math.pow(2, binaryScaleFactor);

  const result: number[] = [];

  if (numberOfBits === 0) {
    for (let i = 0; i < numberOfDataPoints; i++) result.push(referenceValue);
    return result;
  }

  let dataIndex = 0;
  // A bit map applies to this product
  if (bitMapIndicator.code === 0 && bitBuffer !== null) {
    const bitMapData = new Uint8Array(bitBuffer.buffer, bitBuffer.byteOffset, bitBuffer.byteLength);

    for (let i = 0; i < numberOfDataPoints; i++) {
      // GRIB2 Bitmaps are MSB (Most Significant Bit) first
      const byte = bitMapData[i >> 3];
      const bitIsSet = (byte & (0x80 >> (i % 8))) !== 0;

      if (bitIsSet) {
        // Pull from the next available JPEG pixel
        const rawValue = pixels[dataIndex++];
        result.push((referenceValue + rawValue * EE) / DD);
      } else {
        // Set to NaN so the value is ignored upstream
        result.push(NaN);
      }
    }
  } else {
    // No bitmap: 1:1 mapping
    for (const rawValue of pixels) {
      result.push((referenceValue + rawValue * EE) / DD);
    }
  }

  while (result.length < numberOfDataPoints) result.push(NaN);

  return result;
}

export async function pngUnpacking(
  reader: Reader,
  drs: Grib2DataRepresentationSection,
): Promise<number[]> {
  const { numberOfBits, decimalScaleFactor, referenceValue, binaryScaleFactor } =
    drs.dataRepresentation;

  const DD = Math.pow(10, decimalScaleFactor);
  const EE = Math.pow(2, binaryScaleFactor);
  const imageData = new Uint8Array(reader.slice().buffer);
  const decoded = (await imageDecoder(imageData)).data;
  const pixels = new Uint8Array(decoded.buffer);

  const totalBits = pixels.length * 8;
  const count = Math.floor(totalBits / numberOfBits);
  const values: number[] = [];

  let bitPos = 0;

  for (let i = 0; i < count; i++) {
    let acc = 0;

    for (let b = 0; b < numberOfBits; b++) {
      const byteIndex = Math.floor(bitPos / 8);
      const bitOffset = 7 - (bitPos % 8);
      const bit = (pixels[byteIndex] >> bitOffset) & 1;

      acc = (acc << 1) | bit;
      bitPos++;
    }

    values.push((referenceValue + acc * EE) / DD);
  }

  return values;
}

File: rust/readers/grib2/sections/_7/templates.rs
use crate::parsers::{Buffer, image_decoder};
#[cfg(feature = "std")]
use crate::{
    parsers::{BufferReader, Reader},
    readers::{
        _7::{complex_unpacking::*, jpeg2000::*},
        Grib2DataRepresentationSection, Grib2Sections, Grib2Table5_0 as DataType, Grib2Template50,
    },
};
use alloc::vec::Vec;
use libm::pow;

// TODO: case 42: AEC https://github.com/NOAA-EMC/NCEPLIBS-g2c/blob/develop/src/aecunpack.c

/// Decode GRIB2 Template 7.X
/// Converts data Buffer according to data representation section
///
/// ## Parameters
/// - `reader`: The raw data to convert
/// - `sections`: The sections of the GRIB2 message that have been parsed so far
///
/// ## Returns
/// Converted data
pub fn grib2_template_7_decoder(data: &BufferReader, sections: &Grib2Sections) -> Vec<f64> {
    let drs = sections.data_representation.as_ref().unwrap_or_else(|| {
        panic!("Data Representation Section is not defined");
    });
    let data_representation_template = drs.data_representation_template;

    match data_representation_template {
        DataType::GridPointDataSimplePacking => simple_unpacking(data, drs),
        DataType::GridPointDataComplexPacking
        | DataType::GridPointDataComplexPackingAndSpatialDifferencing => {
            complex_unpacking(data, sections)
        }
        DataType::GridPointDataJpeg2000CodeStreamFormat
        | DataType::GridPointDataJpeg2000CodeStreamFormatAndSpatialDifferencing => {
            let bms = sections.bit_map.as_ref().unwrap_or_else(|| {
                panic!("Bit Map Section is not defined");
            });
            if cfg!(not(feature = "std")) {
                unimplemented!("JPEG2000 not implemented yet for no_std. Use the filesystem (std).")
            } else {
                jpeg2000_unpacking_fs(data, drs, bms)
            }
        }
        DataType::GridPointDataPortableNetworkGraphicsPng => png_unpacking(data, drs),
        DataType::SpectralDataSimplePacking => {
            unimplemented!("Spectral simple packing not implemented yet")
        }
        DataType::SpectralDataComplexPacking => {
            unimplemented!("Spectral complex packing not implemented yet")
        }
        _ => panic!("Template 7.{data_representation_template} not defined"),
    }
}

/// # Data Template 7.0 - Grid point data - simple packing
///
/// [Read more...](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_temp7-0.shtml)
///
/// ## Parameters
/// - `reader`: The raw data to convert
/// - `drs`: The data representation section
///
/// ## Returns
///  The converted data
pub fn simple_unpacking(reader: &BufferReader, drs: &Grib2DataRepresentationSection) -> Vec<f64> {
    let Grib2Template50 {
        decimal_scale_factor,
        binary_scale_factor,
        reference_value,
        number_of_bits,
        ..
    } = drs.data_representation.get_simple_packing_template().unwrap_or_else(|| {
        panic!("Simple packing template is not defined");
    });
    let decimal_scale_factor = *decimal_scale_factor as f64;
    let binary_scale_factor = *binary_scale_factor as f64;
    let reference_value = *reference_value as f64;
    let number_of_bits = *number_of_bits;
    let dd = pow(10., decimal_scale_factor);
    let ee = pow(2., binary_scale_factor);

    let data = reader.slice(None, None);
    let total_bits = data.len() * 8;
    let count = total_bits / (number_of_bits as usize);
    let mut values = Vec::with_capacity(count);

    let mut bit_pos = 0;

    for _ in 0..count {
        let mut acc: u32 = 0;

        for _ in 0..number_of_bits {
            let byte_index = bit_pos / 8;
            let bit_offset = 7 - (bit_pos % 8);
            let bit = (data[byte_index] >> bit_offset) & 1;

            acc = (acc << 1) | (bit as u32);
            bit_pos += 1;
        }

        let value = (reference_value + (acc as f64) * ee) / dd;
        values.push(value);
    }

    values
}

fn png_unpacking(reader: &BufferReader, drs: &Grib2DataRepresentationSection) -> Vec<f64> {
    let Grib2Template50 {
        decimal_scale_factor,
        binary_scale_factor,
        reference_value,
        number_of_bits,
        ..
    } = drs.data_representation.get_simple_packing_template().unwrap_or_else(|| {
        panic!("Simple packing template is not defined");
    });
    let decimal_scale_factor = *decimal_scale_factor as f64;
    let binary_scale_factor = *binary_scale_factor as f64;
    let reference_value = *reference_value as f64;
    let number_of_bits = *number_of_bits;
    let dd = pow(10., decimal_scale_factor);
    let ee = pow(2., binary_scale_factor);

    let data = reader.slice(None, None);
    // Parse the image
    let Ok(mut image) = image_decoder(&Buffer::from(data), None) else {
        panic!("Failed to decode image");
    };
    let data = image.data.take();

    let total_bits = data.len() * 8;
    let count = total_bits / (number_of_bits as usize);
    let mut values = Vec::with_capacity(count);

    let mut bit_pos = 0;

    for _ in 0..count {
        let mut acc: u32 = 0;

        for _ in 0..number_of_bits {
            let byte_index = bit_pos / 8;
            let bit_offset = 7 - (bit_pos % 8);
            let bit = (data[byte_index] >> bit_offset) & 1;

            acc = (acc << 1) | (bit as u32);
            bit_pos += 1;
        }

        let value = (reference_value + (acc as f64) * ee) / dd;
        values.push(value);
    }

    values
}
